### PR TITLE
Include fonts from webjars (not source control)

### PR DESCRIPTION
**Pull request by bickelj**
Declare a dependency on a webjar for Source Sans Pro font. The jar includes the font files we need. Use a gradle build task to extract the files from the jar into the resources directory at build time. The `woff` and `woff2` files get included in the resulting theme jar. This way we are not committing binary files to source control.

Issue #7 Theme the user-facing Keycloak pages

### CODE REVIEW

**Review by reefdog — COMMENTED** [review]
This is so great, thanks @bickelj! Much better.

One concern:

The `@font-face` declarations indicate precisely which Unicode range each font file should be used for. (That way, when a 400-weight Cyrillic character is encountered, the corresponding font file is loaded.)

Those `unicode-range` values, though, were generated by our FontSource package at build time, and are specific to the version of the font files at that moment. It's possible that the relevant range will change in a future update to the underlying font file. Our current approach in this PR would update the underlying file, but not the relevant `unicode-range` for each file.

**Does this JAR include any solution for this, that you can tell?** The equivalent behavior would be dynamically generating the updated `@font-face` definitions and bundling them up with the CSS.

File: pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css [hunk 1]
```diff
@@ -2,27 +2,27 @@
...
-@font-face{font-display:swap;font-family:Source Sans Pro;font-style:normal;font-weight:700;src:url(../fonts/source-sans-pro-latin-ext-700-normal.8822d643b28f1ce6bd54.woff2) format("woff2"),url(../fonts/source-sans-pro-all-700-normal.fc594703b14b8a869dfc.woff) format("woff");unicode-range:u+0100-024f,u+0259,u+1e??,u+2020,u+20a0-20ab,u+20ad-20cf,u+2113,u+2c60-2c7f,u+a720-a7ff}
-@font-face{font-display:swap;font-family:Source Sans Pro;font-style:normal;font-weight:700;src:url(../fonts/source-sans-pro-latin-700-normal.a10519031679e736153a.woff2) format("woff2"),url(../fonts/source-sans-pro-all-700-normal.fc594703b14b8a869dfc.woff) format("woff");unicode-range:u+00??,u+0131,u+0152-0153,u+02bb-02bc,u+02c6,u+02da,u+02dc,u+2000-206f,u+2074,u+20ac,u+2122,u+2191,u+2193,u+2212,u+2215,u+feff,u+fffd}
+@font-face{font-display:swap;font-family:Source Sans Pro;font-style:normal;font-weight:400;src:url(../fonts/source-sans-pro-cyrillic-ext-400-normal.woff2) format("woff2"),url(../fonts/source-sans-pro-all-400-normal.woff) format("woff");unicode-range:u+0460-052f,u+1c80-1c88,u+20b4,u+2de0-2dff,u+a640-a69f,u+fe2e-fe2f}
```
Comment (jasonaowen) [inline pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css:5]: Moving what @reefdog [wrote](https://github.com/PhilanthropyDataCommons/auth/pull/15#pullrequestreview-1373784211) to a thread:

> The `@font-face` declarations indicate precisely which Unicode range each font file should be used for. (That way, when a 400-weight Cyrillic character is encountered, the corresponding font file is loaded.)
> 
> Those `unicode-range` values, though, were generated by our FontSource package at build time, and are specific to the version of the font files at that moment. It's possible that the relevant range will change in a future update to the underlying font file. Our current approach in this PR would update the underlying file, but not the relevant `unicode-range` for each file.
> 
> **Does this JAR include any solution for this, that you can tell?** The equivalent behavior would be dynamically generating the updated `@font-face` definitions and bundling them up with the CSS.

Great callout; I hadn't noticed that!

The jar contains everything in the [source package](https://github.com/fontsource/fontsource/tree/main/fonts/google/source-sans-pro). It looks like potentially relevant files are [`index.css`](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/index.css) if that has everything we need, the various `[2-9]00{,-italic}.css` files for specifics weights, or [`mixins.scss`](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/scss/mixins.scss) if we wanted to add SASS to our build process.

@reefdog could we include some of those instead of defining all this CSS directly in `style.css`?

**Review by jasonaowen — COMMENTED** [review]
@bickelj fantastic work! Thanks for knocking this out so quickly. Well done!

File: pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css (hunk 1 above)
Comment (reefdog) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css:5]: Yesss! That would be perfect.

I believe we would want [400.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/400.css), [600.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/600.css), and [700.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/700.css). I don't think we use italics at all, but if we ever do (such as if we decide to allow limited Markdown formatting in some fields), we'll need to grab the `-italic` variants too.

Comment (bickelj) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css:5]: I am not familiar with all this fancy font-wrangling and pretty sure I do not have a firm grasp on what is going on here. I am happy to learn it as I go. But yes, as @jasonaowen says there is more in the jar file. I see, for example, in the directory above `files` a bunch of `css` files, e.g. `200.css` which starts like this:

```css
/* source-sans-pro-cyrillic-ext-200-normal*/
@font-face {
  font-family: 'Source Sans Pro';
  font-style: normal;
  font-display: swap;
  font-weight: 200;
  src: url('./files/source-sans-pro-cyrillic-ext-200-normal.woff2') format('woff2'), url('./files/source-sans-pro-all-200-normal.woff') format('woff');
  unicode-range: U+0460-052F,U+1C80-1C88,U+20B4,U+2DE0-2DFF,U+A640-A69F,U+FE2E-FE2F;
}
/* source-sans-pro-cyrillic-200-normal*/
@font-face {
  font-family: 'Source Sans Pro';
  font-style: normal;
  font-display: swap;
  font-weight: 200;
  src: url('./files/source-sans-pro-cyrillic-200-normal.woff2') format('woff2'), url('./files/source-sans-pro-all-200-normal.woff') format('woff');
  unicode-range: U+0301,U+0400-045F,U+0490-0491,U+04B0-04B1,U+2116;
}
/* ... */
```

Given it is referencing the `files` directory, I think we would probably keep the `files` directory in the build, e.g. `[yada]/resources/fonts/files` rather than only `[yada]/resources/fonts` so that those relative paths could work unchanged.

For a moment I thought that Keycloak only supports one CSS file in themes, but no, it appears [multiple are fine](https://www.keycloak.org/docs/20.0.5/server_development/#add-a-stylesheet-to-a-theme):
> You can add one or more stylesheets to a theme.

In summary, it looks like a lot is possible. As for how much of whose time should be devoted to this, I am not sure.

Comment (jasonaowen) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css:5]: @bickelj oh, great find on the multiple CSS files configuration syntax! I think you have the most robust understanding of how to test these changes; could you please delete the `font-face` lines from `style.css` and include the style sheets @reefdog mentioned:

> I believe we would want [400.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/400.css), [600.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/600.css), and [700.css](https://github.com/fontsource/fontsource/blob/main/fonts/google/source-sans-pro/700.css).

Comment (reefdog) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/resources/css/style.css:5]: > Given it is referencing the `files` directory, I think we would probably keep the `files` directory in the build, e.g. `[yada]/resources/fonts/files` rather than only `[yada]/resources/fonts` so that those relative paths could work unchanged.

Seems sensible.

@bickelj If we can include those three CSS files (and if all the relative references work out), then we can remove all the `@font-face` stuff from our stylesheet and hopefully dust our hands of it!

File: pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties [hunk 2]
```diff
@@ -1,3 +1,3 @@
 parent=keycloak
 import=common/keycloak
-styles=css/style.css
+styles=fonts/source-sans-pro/400.css fonts/source-sans-pro/600.css fonts/source-sans-pro/700.css css/style.css
```
Comment (bickelj) [inline pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties:3]: The lack of a newline is on purpose. It seemed to fix an issue where a non-existent style would be loaded. I guess Keycloak interpreted the newline as another style at `/`.

**Review by reefdog — APPROVED** [review]
This is great @bickelj, thank you for solving and iterating so much, and to @jasonaowen for finding that JAR that let us keep binaries and `@font-face` definitions up to date and out of the repo!

**Review by jasonaowen — APPROVED** [review]
> This is great @bickelj, thank you for solving and iterating so much

Yes, thank you so much, @bickelj!

File: pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties (hunk 2 above)
Comment (jasonaowen) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties:3]: This seems like a bug! I didn't see any relevant issues [upstream](https://github.com/keycloak/keycloak/issues); can you reproduce it with a minimal example and post it, @bickelj?

Comment (bickelj) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties:3]: Agree, I probably won't get to it today, but I can try again sometime next week.

Comment (bickelj) [reply pdc-keycloak-theme/src/main/resources/theme/pdc-keycloak-theme/login/theme.properties:3]: TODO: prioritize attempting to reproduce and submit an issue if it can be reproduced.
